safehttp/header.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"net/http"
+	"net/textproto"
+)
+
+var disallowedHeaders = map[string]bool{"Set-Cookie": true}
+
+// Header represents the key-value pairs in an HTTP header.
+// The keys will be in canonical form, as returned by
+// textproto.CanonicalMIMEHeaderKey.
+type Header struct {
+	wrappedHeader http.Header
+	immutable     map[string]bool
+}
+
+func newHeader(h http.Header) Header {
+	return Header{wrappedHeader: h, immutable: map[string]bool{}}
+}
+
+// MarkImmutable marks the header with the given name as immutable.
+// This header is now read-only. The name is first canonicalized
+// using textproto.CanonicalMIMEHeaderKey.
+func (h Header) MarkImmutable(name string) {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return
+	}
+	h.immutable[name] = true
+}
+
+// Set sets the header with the given name to the given value.
+// The name is first canonicalized using textproto.CanonicalMIMEHeaderKey.
+// If this headers is not immutable, this function removes all other
+// values currently associated with this header before setting the new
+// value. Returns an error when applied on immutable headers.
+func (h Header) Set(name, value string) error {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
```

Comment:
I'm sorry, what is incremental migration?

How would such a switch be implemented? You pass a constant which lives in another package (for easy tracking) as an argument to the constructor of `Header` to disable this check?

Specifically for the `Set-Cookie` header, which is the only one currently in `disallowedHeaders`, it can currently already be set using the `func SetCookie(w ResponseWriter, cookie *Cookie)` function in `net/http`. So migrating to safehttp would just mean switching to using `safehttp.Header.SetCookie()` instead.

Reply:
In this case, use an abstraction like `isCookie` instead of `disallowedHeaders`.

Reply:
It seems from your other comments that a better abstraction here would be `isCookie` than `disallowedHeaders`, so you might disregard this specific comment.

Reply:
I changed from `disallowedHeaders` to just restrict the `Set-Cookie` header in 2fa58a0